### PR TITLE
Handle the volumes created by the CSI driver

### PR DESCRIPTION
**Pull request by ywk253100**
Current VolumeSnapshotter works with volumes allocated by the "in-tree" storage driver. Kubernetes is moving to the CSI driver for Azure and will deprecate the in-tree driver, however we may not be able to get CSI snapshots in Velero to GA before that happens. As an interim solution, we handle volumes created by the CSI driver with the existing VolumeSnapshotter.

Fixes https://github.com/vmware-tanzu/velero/issues/4109

Signed-off-by: Wenkai Yin(尹文开) <yinw@vmware.com>

### CODE REVIEW

**Comment by ywk253100** [comment]
Have done some basic verifications on AKS

File: velero-plugin-for-microsoft-azure/volume_snapshotter.go [hunk 1]
```diff
@@ -396,6 +399,10 @@ func (b *VolumeSnapshotter) GetVolumeID(unstructuredPV runtime.Unstructured) (st
 		return "", errors.WithStack(err)
 	}
 
+	if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == "disk.csi.azure.com" {
+		return strings.TrimPrefix(diskURIRegexp.FindString(pv.Spec.CSI.VolumeHandle), "/Microsoft.Compute/disks/"), nil
```
Comment (reasonerjt) [inline velero-plugin-for-microsoft-azure/volume_snapshotter.go:403]: Should we print some log or return an error if the driver is not "disk.csi.azure.com"

File: velero-plugin-for-microsoft-azure/volume_snapshotter.go [hunk 2]
```diff
@@ -413,12 +420,19 @@ func (b *VolumeSnapshotter) SetVolumeID(unstructuredPV runtime.Unstructured, vol
 		return nil, errors.WithStack(err)
 	}
 
-	if pv.Spec.AzureDisk == nil {
-		return nil, errors.New("spec.azureDisk not found")
-	}
+	if pv.Spec.CSI != nil && pv.Spec.CSI.Driver == "disk.csi.azure.com" {
+		pv.Spec.CSI.VolumeHandle = getComputeResourceName(b.disksSubscription, b.disksResourceGroup, disksResource, volumeID)
```
Comment (reasonerjt) [inline velero-plugin-for-microsoft-azure/volume_snapshotter.go:424]: Should return an error if the driver is not "disk.csi.azure.com"?

**Comment by reasonerjt** [comment]
Could you also test this scenario?
https://github.com/vmware-tanzu/velero/issues/4154

I know this may not be completely related to csi but aws csi work is blocked by such problem, I wanna see if there's a common way to collect the AZ of the PV and put it in item backupper.

**Comment by ywk253100** [comment]
> Could you also test this scenario? [vmware-tanzu/velero#4154](https://github.com/vmware-tanzu/velero/issues/4154)
> 
> I know this may not be completely related to csi but aws csi work is blocked by such problem, I wanna see if there's a common way to collect the AZ of the PV and put it in item backupper.

I can reproduce the issue mentioned in [vmware-tanzu/velero#4154](https://github.com/vmware-tanzu/velero/issues/4154) withe the AKS provisioned in the AZ-supported region

**Comment by ywk253100** [comment]
Tested with PR https://github.com/vmware-tanzu/velero/pull/4350 and works well as expected

File: velero-plugin-for-microsoft-azure/volume_snapshotter.go [hunk 3]
```diff
@@ -413,12 +425,22 @@ func (b *VolumeSnapshotter) SetVolumeID(unstructuredPV runtime.Unstructured, vol
...
 	}
 
-	if pv.Spec.AzureDisk == nil {
-		return nil, errors.New("spec.azureDisk not found")
-	}
+	if pv.Spec.CSI != nil {
+		if pv.Spec.CSI.Driver == diskCSIDriver {
+			pv.Spec.CSI.VolumeHandle = getComputeResourceName(b.disksSubscription, b.disksResourceGroup, disksResource, volumeID)
+			if _, exist := pv.Spec.CSI.VolumeAttributes["csi.storage.k8s.io/pv/name"]; exist {
+				pv.Spec.CSI.VolumeAttributes["csi.storage.k8s.io/pv/name"] = volumeID
```
Comment (reasonerjt) [inline velero-plugin-for-microsoft-azure/volume_snapshotter.go:432]: I don't find too much description about the attribute `csi.storage.k8s.io/pv/name` but it seems the name of the PV not the volume ID of the cloud provider?

**Review by reasonerjt — APPROVED** [review]
lgtm

**Review by zubron — APPROVED** [review]

**Pull request merged**